### PR TITLE
feat : JRD audit functionality updated (RDCC -1112 )

### PR DESCRIPTION
**Pull request by choudsc**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/RDCC-1112

### Change description ###

feat : JRD audit functionality updated (RDCC -1112 ) 

**Does this PR introduce a breaking change?** (check one with "x")

NO

### CODE REVIEW

**Comment by choudsc** [comment]
Adding new PRE

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/util/JrdUtility.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package uk.gov.hmcts.reform.juddata.camel.util;
+
+import java.sql.Timestamp;
+import java.util.HashMap;
+import java.util.Map;
+import javax.transaction.Transactional;
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/util/JrdUtility.java:6]: Rename this class to DataLoadAudit instead of utility

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/route/ParentOrchestrationRoute.java [hunk 2]
```diff
@@ -100,7 +108,11 @@ public void configure() throws Exception {
                         //logging exception in global exception handler
                         onException(Exception.class)
                                 .handled(true)
-                                .process(exceptionProcessor);
+                                .choice().when(header("SchedulerStatus").isEqualTo("PartialSuccess"))
+                                .process(schedulerAuditProcessor)
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/route/ParentOrchestrationRoute.java:112]: can we read the status from constant file?

File: src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java [hunk 3]
```diff
@@ -177,6 +192,67 @@ public void testParentOrchestrationSingleRecord() throws Exception {
...
+        producerTemplate.sendBodyAndHeaders(startRoute, "test JRD orchestration",headers);
+
+        List<Map<String, Object>> judicialUserProfileList = jdbcTemplate.queryForList(sql);
+        assertEquals(judicialUserProfileList.size(), 0);
+
+        List<Map<String, Object>> judicialAppointmentList = jdbcTemplate.queryForList(sqlChild1);
+        assertEquals(judicialAppointmentList.size(), 0);
+
+        List<Map<String, Object>>  dataloadSchedularAudit = jdbcTemplate.queryForList("select *  from dataload_schedular_audit");
+        assertEquals(dataloadSchedularAudit.get(0).get("scheduler_status"), "FAILURE");
```
Comment (shashank-rastogi) [inline src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java:214]: the query should be read from  yaml file and should not be hardcoded here

File: src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java [hunk 4]
```diff
@@ -177,6 +192,67 @@ public void testParentOrchestrationSingleRecord() throws Exception {
...
+        Timestamp timestamp = new Timestamp(System.currentTimeMillis());
+        Map<String, Object> headers = JrdUtility.getSchedulerHeader(schedulerName, timestamp);
+        producerTemplate.sendBodyAndHeaders(startRoute, "test JRD orchestration",headers);
+        List<Map<String, Object>> judicialUserProfileList = jdbcTemplate.queryForList(sql);
+        assertEquals(judicialUserProfileList.size(), 1);
+
+        List<Map<String, Object>> judicialAppointmentList = jdbcTemplate.queryForList(sqlChild1);
+        assertEquals(judicialAppointmentList.size(), 1);
+        List<Map<String, Object>>  dataloadSchedularAudit = jdbcTemplate.queryForList("select *  from dataload_schedular_audit");
+        assertEquals(dataloadSchedularAudit.get(3).get("scheduler_status"), "Success");
```
Comment (shashank-rastogi) [inline src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java:233]: same here.. query shud not be hardcode here

**Comment by choudsc** [comment]
HI Shashank ,

I have made the changes as you revieved.

**Comment by choudsc** [comment]
@abhijitdiwan 
@shashank-rastogi  
@shreedhar-hmcts 
could you please review the  code

File: infrastructure/main.tf [hunk 5]
```diff
@@ -39,3 +39,16 @@ data "azurerm_key_vault_secret" "BLOB_URL_SUFFIX" {
...
+resource "azurerm_resource_group" "rg" {
+  name = "${var.product}-${var.component}-${var.env}"
+  location = "${var.location}"
+  tags {
+    "Deployment Environment" = "${var.env}"
+    "Team Name" = "${var.team_name}"
+    "lastUpdated" = "${timestamp()}"
+  }
+}
+
```
Comment (shashank-rastogi) [inline infrastructure/main.tf:54]: why is this added again? This was removed in last commit by me and should not be added back

File: src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java [hunk 6]
```diff
@@ -4,8 +4,7 @@
 import static org.junit.Assert.assertEquals;
 import static uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.JUDICIAL_USER_PROFILE_ORCHESTRATION;
 
-import java.util.List;
-import java.util.Map;
+import java.util.*;
```
Comment (shashank-rastogi) [inline src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java:7]: dont use * while importing

File: src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java [hunk 7]
```diff
@@ -89,6 +114,8 @@ public void testParentOrchestration() throws Exception {
 
         setSourceData(file);
         camelContext.getGlobalOptions().put(MappingConstants.ORCHESTRATED_ROUTE, JUDICIAL_USER_PROFILE_ORCHESTRATION);
+        camelContext.getGlobalOptions().put(uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.SCHEDULER_START_TIME, uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.getCurrentTimeStamp().toString());
+        camelContext.getGlobalOptions().put(uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.SCHEDULER_NAME, schedulerName);
         parentRoute.startRoute();
```
Comment (shashank-rastogi) [inline src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java:119]: you can move this code to a separate method and add annotation @before to avoid repeating code in every method

File: src/integrationTest/resources/db/testmigration/V1_1__init_tables.sql [hunk 8]
```diff
@@ -101,6 +101,15 @@ CREATE TABLE region_type(
...
 
+CREATE TABLE dataload_schedular_audit(
+  audit_id SERIAL ,
+  scheduler_name varchar(64) NOT NULL,
+  scheduler_start_time timestamp NOT NULL,
+  scheduler_end_time timestamp,
+  scheduler_status varchar(32),
+  CONSTRAINT audit_id PRIMARY KEY (audit_id)
+);
+
```
Comment (shashank-rastogi) [inline src/integrationTest/resources/db/testmigration/V1_1__init_tables.sql:112]: refer to the table ddl from confluence and use it

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ExceptionProcessor.java [hunk 9]
```diff
@@ -7,13 +7,17 @@
...
 @Slf4j
 public class ExceptionProcessor implements Processor {
 
     @Override
     public void process(Exchange exchange) {
         Exception exception = (Exception) exchange.getProperty(EXCEPTION_CAUGHT);
+        if (exception instanceof java.sql.SQLException || exception instanceof org.springframework.dao.DataIntegrityViolationException) {
+            exchange.getContext().getGlobalOptions().put(uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.SCHEDULER_STATUS, uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.FAILURE);
+        }
         log.error("::::exception in route for data processing::::" + exception);
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ExceptionProcessor.java:21]: what happens when any other exception is thrown?

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/SchedulerAuditProcessor.java [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+@Slf4j
+@EnableTransactionManagement
+public class SchedulerAuditProcessor implements Processor {
+
+    @Autowired
+    DataLoadAudit dataLoadAudit;
+
+    @Value("${audit-enable}")
+    Boolean auditEnablel;
+
```
Comment (shashank-rastogi) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/SchedulerAuditProcessor.java:22]: spelling in wrong.. correct it

File: src/test/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAuditTest.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    @Before
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+    }
+
+    @Test
+    public void testSchedularAuditUpdate() throws Exception {
+
+        final String schedulerName = "judicial_main_scheduler";
+        final String schedulerStatus = "Test";
```
Comment (shashank-rastogi) [inline src/test/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAuditTest.java:48]: do not hardcode scheduler name here.. it shud be picked from application yanl file

File: src/test/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAuditTest.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+        Mockito.when(exchange.getContext().getGlobalOptions()).thenReturn(globalOptions);
+        Message messageMock = Mockito.mock(Message.class);
+
+        Mockito.when(camelContext.getGlobalOptions()).thenReturn(globalOptions);
+        Timestamp schedulerStartTime = MappingConstants.getCurrentTimeStamp();
+        Mockito.when(mockJdbcTemplate.update(schedulerInsertSql, schedulerName, schedulerStartTime, new Timestamp(System.currentTimeMillis()), schedulerStatus)).thenReturn(0);
+
+        dataLoadAuditUnderTest.schedularAuditUpdate(exchange);
+    }
+}
```
Comment (shashank-rastogi) [inline src/test/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAuditTest.java:64]: where is the assert statement?

File: src/test/java/uk/gov/hmcts/reform/juddata/camel/processor/SchedulerAuditProcessorTest.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+    @Test
+    public void testProcess() throws Exception {
+        // Setup
+        final Exchange exchange = new DefaultExchange(context);
+        Message message = Mockito.mock(Message.class);
+
+        schedulerAuditProcessorUnderTest.auditEnablel = Boolean.TRUE;
+        // Run the test
+        schedulerAuditProcessorUnderTest.process(exchange);
+
```
Comment (shashank-rastogi) [inline src/test/java/uk/gov/hmcts/reform/juddata/camel/processor/SchedulerAuditProcessorTest.java:32]: how are we testing false condition ?

File: src/main/resources/application-jrd-leaf-router.yaml [hunk 14]
```diff
@@ -1,5 +1,8 @@
 start-leaf-route: direct:leaf
 leaf-route-names: judicial-role-type,judicial-contract-type,judicial-region-type,judicial-base-location_type
+Scheduler-insert-sql:  insert into schedular_audit (scheduler_Id,scheduler_Name,scheduler_Start_Time,scheduler_End_Time,scheduler_Status)
+  values(DEFAULT,?,?,?,?)
```
[the remark this answers is not in the document]
Comment (shashank-rastogi) [reply src/main/resources/application-jrd-leaf-router.yaml:4]: is this addressed?

File: src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java (hunk 3 above)
Comment (shashank-rastogi) [reply src/integrationTest/java/uk/gov/hmcts/reform/juddata/cameltest/ParentOrchestrationRouteTest.java]: is this addressed?

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ExceptionProcessor.java [hunk 15]
```diff
@@ -7,13 +7,17 @@
...
+
 @Component
 @Slf4j
 public class ExceptionProcessor implements Processor {
 
     @Override
     public void process(Exchange exchange) {
         Exception exception = (Exception) exchange.getProperty(EXCEPTION_CAUGHT);
+        if (exception instanceof java.sql.SQLException || exception instanceof org.springframework.dao.DataIntegrityViolationException) {
+            exchange.getContext().getGlobalOptions().put(uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.SCHEDULER_STATUS, uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.FAILURE);
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ExceptionProcessor.java:19]: use static imports for MappingCosntants

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/SchedulerAuditProcessor.java [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+@Component
+@Slf4j
+@EnableTransactionManagement
+public class SchedulerAuditProcessor implements Processor {
+
+    @Autowired
+    DataLoadAudit dataLoadAudit;
+
+    @Value("${audit-enable}")
+    Boolean auditEnablel;
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/SchedulerAuditProcessor.java:21]: can be renamed to auditEnabled

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAudit.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+    private String schedulerInsertSql;
+
+    @Transactional(Transactional.TxType.REQUIRES_NEW)
+    public void schedularAuditUpdate(final Exchange exchange) {
+
+        Map<String, String> globalOptions = exchange.getContext().getGlobalOptions();
+        Timestamp schedulerStartTime = Timestamp.valueOf(globalOptions.get(MappingConstants.SCHEDULER_START_TIME));
+        String schedulerName = globalOptions.get(MappingConstants.SCHEDULER_NAME);
+        String schedulerStatus = globalOptions.get(MappingConstants.SCHEDULER_STATUS);
+        if (Strings.isNullOrEmpty(schedulerStatus) || schedulerStatus.equalsIgnoreCase(MappingConstants.PARTIAL_SUCCESS)) {
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAudit.java:33]: static import all MappingConstants

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAudit.java [hunk 18]
```diff
@@ -0,0 +1,42 @@
...
+public class DataLoadAudit {
+
+    @Autowired
+    private JdbcTemplate jdbcTemplate;
+
+    @Value("${Scheduler-insert-sql}")
+    private String schedulerInsertSql;
+
+    @Transactional(Transactional.TxType.REQUIRES_NEW)
+    public void schedularAuditUpdate(final Exchange exchange) {
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAudit.java:27]: rename method name to processAudit() or something meaningful

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAudit.java [hunk 19]
```diff
@@ -0,0 +1,42 @@
...
+
+        Map<String, String> globalOptions = exchange.getContext().getGlobalOptions();
+        Timestamp schedulerStartTime = Timestamp.valueOf(globalOptions.get(MappingConstants.SCHEDULER_START_TIME));
+        String schedulerName = globalOptions.get(MappingConstants.SCHEDULER_NAME);
+        String schedulerStatus = globalOptions.get(MappingConstants.SCHEDULER_STATUS);
+        if (Strings.isNullOrEmpty(schedulerStatus) || schedulerStatus.equalsIgnoreCase(MappingConstants.PARTIAL_SUCCESS)) {
+            if (Strings.isNullOrEmpty(schedulerStatus)) {
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAudit.java:34]: Strings.isNullOrEmpty(schedulerStatus) is repeated . Also this conditions can be schedulerStatus is null then SUCCESS else PARTIAL_SUCCESS

File: src/main/java/uk/gov/hmcts/reform/juddata/scheduler/JrdLeafDataLoadStarter.java [hunk 20]
```diff
@@ -29,12 +29,22 @@
...
+
     @PostConstruct
     public void postConstruct() throws Exception {
         camelContext.start();
+        setGlobalOption();
         leafTableRoutes.startRoute();
     }
 
+    public void setGlobalOption() {
+        camelContext.getGlobalOptions().put(uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.ORCHESTRATED_ROUTE, uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.JUDICIAL_USER_PROFILE_ORCHESTRATION);
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/scheduler/JrdLeafDataLoadStarter.java:43]: get this camelContext.getGlobalOptions() in one variable and then use. Also use static constants for  MappingConstants

File: src/main/java/uk/gov/hmcts/reform/juddata/scheduler/JrdUserProfileDataLoadScheduler.java [hunk 21]
```diff
@@ -28,15 +27,24 @@
...
     @PostConstruct
     public void postConstruct() throws Exception {
         camelContext.start();
-        camelContext.getGlobalOptions().put(MappingConstants.ORCHESTRATED_ROUTE, JUDICIAL_USER_PROFILE_ORCHESTRATION);
+        setGlobalOption();
         parentOrchestrationRoute.startRoute();
     }
 
+    private void setGlobalOption() {
+        camelContext.getGlobalOptions().put(uk.gov.hmcts.reform.juddata.camel.util.MappingConstants.ORCHESTRATED_ROUTE, JUDICIAL_USER_PROFILE_ORCHESTRATION);
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/scheduler/JrdUserProfileDataLoadScheduler.java:41]: this is same code which is used in both scheduler. Can be put in util class and reused.

File: src/main/resources/application.yaml [hunk 22]
```diff
@@ -60,13 +60,13 @@ spring:
...
+  camel-route-config: 0 0/3 * * * ?
+  camel-leaf-router-config: 0 */1 * * * ?
 
 azure:
   storage:
-    account-name: ${ACCOUNT_NAME:ACCOUNT_NAME}
-    account-key: ${ACCOUNT_KEY:ACCOUNT_KEY}
-    container-name: ${CONTAINER_NAME:CONTAINER_NAME}
+    account-name: ${ACCOUNT_NAME:rddemo}
+    account-key: ${ACCOUNT_KEY:Vjmr3Yk0DUrygumP7FKJ2eUiOyhWvZ5XczDMk13J+m9vtk2IOXEcOlfjlcTaHDb5mHwAsq7qOYf3GtxEo/uFIA==}
```
Comment (shreedhar-hmcts) [inline src/main/resources/application.yaml:69]: revert these credentials to original

File: src/test/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAuditTest.java (hunk 12 above)
Comment (shreedhar-hmcts) [reply src/test/java/uk/gov/hmcts/reform/juddata/camel/util/DataLoadAuditTest.java:64]: check if you can use verify if mock is used

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/route/LeafTableRoute.java [hunk 23]
```diff
@@ -57,12 +59,18 @@
...
+    @Autowired
+    SchedulerAuditProcessor schedulerAuditProcessor;
+
     @Autowired
     CamelContext camelContext;
 
     @SuppressWarnings("unchecked")
     @Transactional
-    public void startRoute() throws FailedToCreateRouteException {
+    public void startRoute() throws Exception {
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/route/LeafTableRoute.java:73]: might be will give SONAR exception, since we can't use Generic exception

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/SchedulerAuditProcessor.java [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+     * Processes the message exchange.
+     *
+     * @param exchange the message exchange
+     * @throws Exception if an internal processing error has occurred.
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public void process(Exchange exchange) throws Exception {
+        if (auditEnablel.booleanValue()) {
+            dataLoadAudit.schedularAuditUpdate(exchange);
```
Comment (shreedhar-hmcts) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/SchedulerAuditProcessor.java:33]: correct spelling and rename method to : auditUpdate or some meaningful name

File: src/integrationTest/resources/application-integration.yml [hunk 25]
```diff
@@ -11,6 +11,10 @@ archival-file-names: test.csv,judicial_userprofile.csv
 archival-cred: noop=true
 archival-date-format: dd-MM-yyyy--HH-mm
 file-read-time-out: 2000
+Scheduler-insert-sql:  insert into dataload_schedular_audit (audit_id,scheduler_name,scheduler_start_time,scheduler_end_time,scheduler_status)
+  values(DEFAULT,?,?,?,?)
```
Comment (abhijitdiwan) [inline src/integrationTest/resources/application-integration.yml:15]: Default not needed.

File: src/integrationTest/resources/application-integration.yml [hunk 26]
```diff
@@ -11,6 +11,10 @@ archival-file-names: test.csv,judicial_userprofile.csv
 archival-cred: noop=true
 archival-date-format: dd-MM-yyyy--HH-mm
 file-read-time-out: 2000
+Scheduler-insert-sql:  insert into dataload_schedular_audit (audit_id,scheduler_name,scheduler_start_time,scheduler_end_time,scheduler_status)
```
Comment (abhijitdiwan) [inline src/integrationTest/resources/application-integration.yml:14]: case Scheduler

File: src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ExceptionProcessor.java [hunk 27]
```diff
@@ -7,13 +7,17 @@
...
 
+
 @Component
 @Slf4j
 public class ExceptionProcessor implements Processor {
 
     @Override
     public void process(Exchange exchange) {
         Exception exception = (Exception) exchange.getProperty(EXCEPTION_CAUGHT);
+        if (exception instanceof java.sql.SQLException || exception instanceof org.springframework.dao.DataIntegrityViolationException) {
```
Comment (abhijitdiwan) [inline src/main/java/uk/gov/hmcts/reform/juddata/camel/processor/ExceptionProcessor.java:18]: please refactor.
